packages/storybook/src/stories/grid-rows.stories.tsx

```diff
@@ -159,3 +165,212 @@ export function ChangeRowsAndColumns() {
+      editRowState[selectedCell[0]] = editRowState[selectedCell[0]]
+        ? editRowState[selectedCell[0]]
+        : {};
+      editRowState[selectedCell[0]][selectedCell[1]] = true;
+
+      return { ...state, ...editRowState };
+    });
+  }, [selectedCell]);
+
+  const onCellClick = React.useCallback((params: CellParams) => {
+    setSelectedCell([params.row.id!.toString(), params.field]);
+    setIsEditable(!!params.isEditable);
+  }, []);
+
+  const onCellDoubleClick = React.useCallback(
+    (params: CellParams) => {
+      if (params.isEditable) {
+        apiRef.current.setCellMode(params.row.id!.toString(), params.field, 'edit');
+      }
+    },
+    [apiRef],
+  );
+
+  const isCellEditable = React.useCallback((params: CellParams) => params.row.id !== 0, []);
+
+  const onEditCellValueChange = React.useCallback(({ update }) => {
+    if (update.email) {
+      const isValid = validateEmail(update.email);
+      // how do we feedback users?
+      // how do we prevent committing?
+      const newState = {};
+      newState[update.id] = { email: { value: update.email, error: !isValid } };
+      setEditRowsModel((state) => ({ ...state, ...newState }));
+    }
+  }, []);
+
+  const onEditCellValueChangeCommitted = React.useCallback(
+    ({ update, api }) => {
+      const field = Object.keys(update).find((key) => key !== 'id')!;
+      if (update.email) {
+        const newState = {};
+        const componentProps = {
+          InputProps: { endAdornment: <LoadIcon /> },
+        };
+        newState[update.id] = {};
+        newState[update.id][field] = { value: update.email, ...componentProps };
+        setEditRowsModel((state) => ({ ...state, ...newState }));
+        setTimeout(() => {
+          apiRef.current.updateRows([update]);
+          apiRef.current.setCellMode(update.id, field, 'view');
+        }, 2000);
+      } else if (update.fullname) {
+        const [firstname, lastname] = update.fullname.split(' ');
+        apiRef.current.updateRows([{ id: update.id, firstname, lastname }]);
+        apiRef.current.setCellMode(update.id, field, 'view');
+      } else {
+        apiRef.current.updateRows([update]);
+        apiRef.current.setCellMode(update.id, field, 'view');
+      }
+    },
+    [apiRef],
+  );
+
+  return (
+    <React.Fragment>
+      Green cells are editable! Click + EDIT or Double click
+      <div style={{ display: 'flex', justifyContent: 'space-around' }}>
+        <Button color="primary" onClick={editRow} disabled={!isEditable}>
+          Edit
+        </Button>
+      </div>
+      <div className="grid-container">
+        <XGrid
+          className={classes.root}
+          {...baselineEditProps}
+          apiRef={apiRef}
+          onCellClick={onCellClick}
+          onCellDoubleClick={onCellDoubleClick}
+          isCellEditable={isCellEditable}
+          onEditCellValueChange={onEditCellValueChange}
+          onEditCellValueChangeCommitted={onEditCellValueChangeCommitted}
```

Comment:
I don't understand this API. Why don't we have a single callback in the root component and 2 callbacks at the renderEditCell level? In another word, why do we need to rerendering the whole grid at each keystroke or even save the state outside of the cell itself?

Reply:
- allows to prepopulate the state with a value
- we pass a default renderEditCell, this allows users to hook into it and validate the new value without overriding the whole component